Node deployment enhancements

eduardostalinho opened the pull request:
This PR adds enchancements for the *node deployment* resource.
- validate node version against API endpoint `/upgrades/node/`
- validate node deployment version against cluster version: node version should not be greater than cluster's current version.
- validate if node is being deployed to the same cloud provider as the cluster.

### TODO:
- [x] fix acceptance tests for current valid versions.
- [ ] add acceptance tests with plan only to test validation errors.
- [ ] add `dynamic_config` field to `node_deployment.spec`.
 
PS: I'm not adding to kubermatic/terraform-provider-kubermatic because my branch is up-to date with this repository.

eduardostalinho committed:
Validate node deployment version against API.

Check against the endpoint that provides available versions for node
deployments on CustomizeDiff.

- CustomizeDiff was used directly on `"version"` nested resource of node
deployment template.
- Validation was made on customizediff to bring access to the api
resources.

Signed-off-by: Eduardo Carvalho <eduardooc.86@gmail.com>

eduardostalinho committed:
Validate if node version against cluster version.

Node deployment versions can't be greater that cluster version.
Check is made against cluster endpoint.

Signed-off-by: Eduardo Carvalho <eduardooc.86@gmail.com>

eduardostalinho committed:
Ignore customdiff if version is not set.

Signed-off-by: Eduardo Carvalho <eduardooc.86@gmail.com>

eduardostalinho committed:
Extract cluster request for node deployment.

Signed-off-by: Eduardo Carvalho <eduardooc.86@gmail.com>

eduardostalinho committed:
Validate node spec against cluster.

Use only one request to cluster to
validate version and provider spec.

Signed-off-by: Eduardo Carvalho <eduardooc.86@gmail.com>

furkhat commented on kubermatic/validation_node_deployment.go:
```diff
@@ -0,0 +1,141 @@
+		versionType := "kubernetes"
+		p := versions.NewGetNodeUpgradesParams()
+		p.SetType(&versionType)
+		p.SetControlPlaneVersion(&version)
```
use cluster version as control plane version

furkhat commented on kubermatic/validation_node_deployment.go:
```diff
@@ -0,0 +1,141 @@
+	case c.Spec.Cloud.Aws != nil:
+		return "aws", nil
+	case c.Spec.Cloud.Openstack != nil:
+		return "openstack", nil
```
add azure

furkhat reviewed (changes requested):
Please add acceptance tests with PlanOnly=true and ExpectError=<regex> fields of `resource.TestCase` to check validation

eduardostalinho committed:
Fix: Use control version to get version for node.

Signed-off-by: Eduardo Carvalho <eduardooc.86@gmail.com>

eduardostalinho committed:
Add azure as available cloud provider.

Signed-off-by: Eduardo Carvalho <eduardooc.86@gmail.com>

Commit:
Add acceptance tests for node deployment validations.

Since the validations are run against an exisitin cluster if it exists,
a new variable must be set with KUBERMATIC_EXISTING_CLUSTER_ID with the format
<PROJECT_ID>:<SEED_DC>:<CLUSTER_ID>

Signed-off-by: Eduardo Carvalho <eduardooc.86@gmail.com>

Commit:
Add dynamic_config field to node deployment.

Signed-off-by: Eduardo Carvalho <eduardooc.86@gmail.com>
